File: python-sdk/axon/registry.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol

from axon.errors import UnsupportedLLMProviderError


class Matcher(Protocol):
    def __call__(self, client: Any) -> bool: ...


class Patcher(Protocol):
    def __call__(self, client: Any, *, axon: Any) -> None: ...


@dataclass(frozen=True, slots=True)
class ClientRegistration:
    matcher: Matcher
    patcher: Patcher


_client_registrations: list[ClientRegistration] = []
_providers_loaded = False


def register_client(*, matcher: Matcher, patcher: Patcher) -> None:
    _client_registrations.append(ClientRegistration(matcher=matcher, patcher=patcher))


def _ensure_providers_loaded() -> None:
    global _providers_loaded
    if _providers_loaded:
        return
    import axon.providers  # noqa: F401

    _providers_loaded = True


def patch_client(*, axon: Any, client: Any) -> None:
    _ensure_providers_loaded()
    for reg in _client_registrations:
        if reg.matcher(client):
            reg.patcher(client, axon=axon)
            return

    provider = f"{type(client).__module__}.{type(client).__name__}"
    raise UnsupportedLLMProviderError(provider)

File: ts-sdk/src/registry.ts
import { UnsupportedLLMProviderError } from "./errors.js";

export type Matcher = (client: unknown) => boolean;
export type Patcher = (client: any, axon: any) => void;

interface ClientRegistration {
  matcher: Matcher;
  patcher: Patcher;
}

const registrations: ClientRegistration[] = [];
let providersLoaded = false;

export function registerClient(matcher: Matcher, patcher: Patcher): void {
  registrations.push({ matcher, patcher });
}

async function ensureProvidersLoaded(): Promise<void> {
  if (providersLoaded) return;
  await import("./providers/index.js");
  providersLoaded = true;
}

export async function patchClient(axon: any, client: any): Promise<void> {
  await ensureProvidersLoaded();

  for (const reg of registrations) {
    if (reg.matcher(client)) {
      reg.patcher(client, axon);
      return;
    }
  }

  const provider = client?.constructor?.name ? String(client.constructor.name) : typeof client;
  throw new UnsupportedLLMProviderError(provider);
}
